clang/runtime/dpct-rt/include/sparse_utils.hpp

```diff
@@ -743,6 +767,316 @@ inline void spmm(sycl::queue queue, oneapi::mkl::transpose trans_a,
+  auto alpha_value =
+      dpct::detail::get_value(reinterpret_cast<const T *>(alpha), queue);
+  if (alpha_value != T(1.0f)) {
+    throw std::runtime_error(
+        "oneapi::mkl::sparse::trsv only supports alpha == 1.");
+  }
+}
+} // namespace detail
+
+/// Performs internal optimizations for spsv by analyzing the provided matrix
+/// structure and operation parameters.
+/// \param [in] queue The queue where the routine should be executed. It must
+/// have the in_order property when using the USM mode.
+/// \param [in] trans_a Specifies operation on input matrix a.
+/// \param [in] a Specifies the sparse matrix a.
+inline void spsv_optimize(sycl::queue queue, oneapi::mkl::transpose trans_a,
+                          sparse_matrix_desc_t a) {
+  if (!a->get_uplo() || !a->get_diag()) {
+    throw std::runtime_error("oneapi::mkl::sparse::trsv needs uplo and diag "
+                             "attributes to be specified.");
+  }
+  oneapi::mkl::sparse::optimize_trsv(
+      queue, a->get_uplo().value(), oneapi::mkl::transpose::nontrans,
+      a->get_diag().value(), a->get_matrix_handle());
+}
+
+/// Solves a system of linear equations for a sparse matrix.
+/// \param [in] queue The queue where the routine should be executed. It must
+/// have the in_order property when using the USM mode.
+/// \param [in] trans_a Specifies operation on input matrix a.
+/// \param [in] alpha Specifies the scalar alpha.
+/// \param [in] a Specifies the sparse matrix a.
+/// \param [in] x Specifies the dense vector x.
+/// \param [in, out] y Specifies the dense vector y.
+/// \param [in] data_type Specifies the data type of \param a, \param x and
+/// \param y .
+inline void spsv(sycl::queue queue, oneapi::mkl::transpose trans_a,
+                 const void *alpha, sparse_matrix_desc_t a,
+                 std::shared_ptr<dense_vector_desc> x,
+                 std::shared_ptr<dense_vector_desc> y,
+                 library_data_t data_type) {
+  switch (data_type) {
+  case library_data_t::real_float:
+    detail::check_alpha_value<float>(alpha, queue);
+    break;
+  case library_data_t::real_double:
+    detail::check_alpha_value<double>(alpha, queue);
+    break;
+  case library_data_t::complex_float:
+    detail::check_alpha_value<std::complex<float>>(alpha, queue);
+    break;
+  case library_data_t::complex_double:
+    detail::check_alpha_value<std::complex<double>>(alpha, queue);
+    break;
+  default:
+    throw std::runtime_error("Unsupported data type.");
+  }
+
+  if (!a->get_uplo() || !a->get_diag()) {
+    throw std::runtime_error("oneapi::mkl::sparse::trsv needs uplo and diag "
+                             "attributes to be specified.");
+  }
+  oneapi::mkl::uplo uplo = a->get_uplo().value();
+  oneapi::mkl::diag diag = a->get_diag().value();
+
+  std::uint64_t key = dpct::detail::get_type_combination_id(
+      a->get_value_type(), x->get_value_type(), y->get_value_type());
+  switch (key) {
```

Comment:
oneMKL Sparse BLAS doesn't support mixed precision. You need to throw an exception if data type of `alpha` does not match these data types (`a`, `x`, `y`).

Reply:
Yes. I noticed this. But I think it is unnecessary for `sparse::trsv` because `sparse::trsv` does not need to use `alpha`, it only support `alpha == 1`. So I think we can ignore alpha's type here.
For other APIs which accept alpha/beta, I will throw the exception if data type mismatch.